test/common/testlib.py

```diff
@@ -777,72 +834,191 @@ def copy_cores(self, title, label=None):
 def jsquote(str):
     return json.dumps(str)
 
-# See phantom-driver for the methods that are defined
-class Phantom:
-    def __init__(self, lang=None):
+class CDP:
+    def __init__(self, lang=None, headless=True):
         self.lang = lang
         self.timeout = 60
         self.valid = False
+        self.headless = headless
         self._driver = None
+        self._browser = None
+        self._browser_home = None
+        self._cdp_port_lockfile = None
 
     def __getattr__(self, name):
         if not name.startswith("_"):
-            return lambda *args: self._invoke(name, *args)
+            return lambda **kwargs: self._invoke(name, kwargs)
         raise AttributeError
 
-    def _invoke(self, name, *args):
+    def _invoke(self, name, kwargs):
+        trace = opts.trace and not kwargs.get("no_trace", False)
+        try:
+            del kwargs["no_trace"]
+        except KeyError:
+            pass
+
+        cmd = name.replace("_", ".") + "(" + json.dumps(kwargs) + ")"
+
+        # frame support for Runtime.evaluate(): map frame name to
+        # executionContextId and insert into argument object; this must not be quoted
+        # see "Frame tracking" in cdp-driver.js for how this works
+        if name == 'Runtime_evaluate' and self.cur_frame:
+            cmd = "%s, contextId: getFrameExecId(%s)%s" % (cmd[:-2], jsquote(self.cur_frame), cmd[-2:])
+
+        if trace:
+            print("-> " + kwargs.get('trace', cmd))
+
+        # avoid having to write the "client." prefix everywhere
+        if cmd.startswith("Page.") or cmd.startswith("Runtime.") or cmd.startswith("Input.") or \
+               cmd.startswith("DOM.") or cmd.startswith("Network."):
+            cmd = "client." + cmd
+        res = self.command(cmd)
+        if trace:
+            if "result" in res:
+                print("<- " + repr(res["result"]))
+            else:
+                print("<- " + repr(res))
+        return res
+
+    def command(self, cmd):
         if not self._driver:
             self.start()
-        if opts.trace:
-            print("-> {0}({1})".format(name, repr(args)[1:-2]))
-        line = json.dumps({
-            "cmd": name,
-            "args": args,
-            "timeout": self.timeout * 1000
-        }).replace("\n", " ") + "\n"
-        self._driver.stdin.write(line)
+        self._driver.stdin.write(cmd + "\n")
         line = self._driver.stdout.readline()
         if not line:
             self.kill()
-            raise Error("PhantomJS or driver broken")
+            raise Error("CDP broken")
+        if not line.startswith("success "):
+            if opts.trace:
+                print("<- raise " + line)
+            raise Error(line)
```

Comment:
Could we make the driver output a single line of json? I know this is not a lot of parsing, but it seems weird to me to support `fail <string>` and `success <json result>` when we could just have a single json blob with either a `result` or `error` key.

Reply:
Done.